csharp/src/Ice/Communicator.cs

```diff
@@ -1364,5 +1273,106 @@ private void AddAllAdminFacets()
 
             return null;
         }
+
+        private async Task PerformDisposeAsync()
+        {
+            lock (_mutex)
+            {
+                Debug.Assert(_state < StateDestroyInProgress);
+                _state = StateDestroyInProgress;
+            }
+
+            // Cancel operations that are waiting and using the communicator's cancellation token
+            _cancellationTokenSource.Cancel();
+
+            // Shutdown and destroy all the incoming and outgoing Ice connections and wait for the connections to be
+            // finished.
+            _ = ShutdownAsync();
+
+            _outgoingConnectionFactory?.DestroyAsync();
```

Comment:
Do we need to call shutdown here? Can't we just Dispose of the OAs without Deactivating them? Instead of calling twice DestroyAsync on the outgoing connection factory, we could also keep track of the returned task and await on it later, it would be a bit cleaner instead of ignore the returned task.

Reply:
Reworked this a bit , but I keep the call to ShutdownAsync, as it is required for WhatForShutdownAsync to return